### ISSUE TITLE
Unable to execute multiple requests in parallel through sam local start-api

### ISSUE THREAD

**Issue by jdeon-classy** [awslabs/aws-sam-cli#301]
**Template:**
AWSTemplateFormatVersion: '2010-09-09'
Transform: 'AWS::Serverless-2016-10-31'
Description: An AWS Serverless Specification template describing your function.
Resources:
  CheckoutLambda:
    Type: "AWS::Serverless::Function"
    Properties:
      Handler: "CheckoutLambda/index.handler"
      Role: redacted
      Runtime: "nodejs6.10"
      Timeout: 300
      Environment:
        Variables:
          ENV: int
      Events:
        CheckoutApi:
          Type: Api
          Properties:
            Path: '/checkout/donate'
            Method: post
**Executing two posts in parallel:**
`curl -X POST   http://127.0.0.1:3000/checkout/donate   -H 'cache-control: no-cache'   -H 'content-type: application/json'   -H 'postman-token: 9f48e8af-a291-c6c6-6e9f-cc60a358a872'   -d '5000' & curl -X POST   http://127.0.0.1:3000/checkout/donate   -H 'cache-control: no-cache'   -H 'content-type: application/json'   -H 'postman-token: 9f48e8af-a291-c6c6-6e9f-cc60a358a872'   -d '4000'`
**Output from aws sam cli:** 
`2018/02/12 14:42:53 Invoking CheckoutLambda/index.handler (nodejs6.10)
2018/02/12 14:42:53 Invoking CheckoutLambda/index.handler (nodejs6.10)
START RequestId: 3c1c1ce5-fb4a-13be-4400-2fc7df6fed59 Version: $LATEST
2018-02-12T22:42:57.276Z	3c1c1ce5-fb4a-13be-4400-2fc7df6fed59	4000
START RequestId: 6d07c29d-da2e-14cc-2d1e-8996b0fccdcf Version: $LATEST
2018-02-12T22:42:57.539Z	6d07c29d-da2e-14cc-2d1e-8996b0fccdcf	5000
END RequestId: 3c1c1ce5-fb4a-13be-4400-2fc7df6fed59
REPORT RequestId: 3c1c1ce5-fb4a-13be-4400-2fc7df6fed59	Duration: 4033.19 ms	Billed Duration: 4100 ms	Memory Size: 0 MB	Max Memory Used: 31 MB	
2018/02/12 14:42:59 Function returned an invalid response (must include one of: body, headers or statusCode in the response object): unexpected end of JSON input`

Another thing to note is that the function then later times out:
`2018/02/12 14:26:05 Function CheckoutLambda/index.handler timed out after 300 seconds`

**handler(index.js)**
`function sleepFor( sleepDuration ){
    var now = new Date().getTime();
    while(new Date().getTime() < now + sleepDuration){ /* do nothing */ }
}
function formatResponse (statusCode, body) {
  return {
    statusCode: statusCode,
    body: JSON.stringify(body)
  };
}
exports.handler = (event, context, callback) => {
  console.log(event.body);
  sleepFor(JSON.parse(event.body));

  callback(null, formatResponse(200,'test'));
};`
This is a contrived example that I could share. But I'm running into this problem when trying to create integration tests for my lambda function. Ideally I'd like to be able to execute many tests in parallel against the locally spun-up api so that it mirrors the actual api. Thanks for anyone who looks into this! :)

PS: Sometimes it works as expected. Race condition?

**Comment by brownjava** [comment]
The code is storing per-request information in a context that spans multiple requests, so there is weird/broken behavior if you issue multiple requests at the same time.  I've put together a cursory PR (sorry, haven't written much go) at:

https://github.com/awslabs/aws-sam-local/pull/304

**Comment by araker** [comment]
Would be really nice to have this functionality, or a queue system to handle multiple requests.

**Comment by hobotroid** [comment]
Yeah it's really difficult to develop and test an API locally when it can only handle 1 request at a time. That, coupled with the fact that each requests takes 4-8 seconds to complete (container has to spin up for every request), makes the whole process painful.

**Comment by jfuss** [comment]
Parallel Requests are now completely supported as of v.0.3.0. We even have [integ tests](https://github.com/awslabs/aws-sam-cli/blob/develop/tests/integration/local/start_api/test_start_api.py#L8) that exercise this :)

Closing as this was addressed.

**Comment by chetan-nandrajog** [comment]
@jfuss : I am still unable to execute multiple request using `sam local start-api` 
Using sam version `SAM CLI, version 0.40.0`
Whenever i try to use concurrent api call i get `503- bad gateway`

**Comment by ben8p** [comment]
@chetan-nandrajog I am experimenting the same. do you use `--docker-network host` ?
Because that's what is causing the issue for me.  
Underlying error: `listen tcp :9001: bind: address already in use`

**Comment by chetan-nandrajog** [comment]
@ben8p that is correct. i are using `--docker-network=host` .  Please let me know if you find a solution for it.

**Comment by ben8p** [comment]
@chetan-nandrajog I will...
All Linux users must have the same issue.

**Comment by seawatts** [comment]
I'm also seeing this same error when running with `--docker-network host`

**Comment by ben8p** [comment]
Eventually I workarounded my problem by using a custom docker network instead of the host.

**Comment by akomiqaia** [comment]
hi @ben8p. I started using a custom docker network. but every time I make a request from my front-end it tries to pull the docker images all the time which take around 10 seconds each time.  The way I created the docker network was by running `docker network create lambda-local` and then tried to use `sam local start-api --docker-network lambda-local`. do you need to add extra flags on sam invocations or are there ways to make simultaneous requests that won't take too long?

**Comment by ben8p** [comment]
@akomiqaia 
Here is the command I use:
`sam local start-api --docker-network=my-docker-network --host 0.0.0.0 --port 3001 --debug --template-file ./my.template.json --docker-volume-basedir ./build`

**Comment by liesislukas** [comment]
to avoid re-building of container for each request you can use `--warm-containers LAZY` flag 

`sam local start-api --warm-containers LAZY`

![image](https://user-images.githubusercontent.com/2733862/144033954-92216b9f-eba5-4233-b6f2-abcd195cddc7.png)

**Comment by Pipas** [comment]
I'm using @liesislukas hack solution for faster local development but would be nice if warm containers would work with parallel requests out of the box

**Comment by CynanX** [comment]
I do not use `--docker-network host`, yet am experiencing the same issue. Doing multiple calls to a Python lambda fails.

It's suggested that this was fixed but it appears not to be the case.

There seems to be some suggesting that it's fixed, but also that it's not the intended use case of SAM and it won't be fixed.... which is it?

**Comment by edgejohn** [comment]
Hello, I might encounter this error as well. If all the lambda functions can be run simultaneously and work as a local API, that would be great for local development.
